src/vanilla.ts

```diff
@@ -1,153 +1,179 @@
+  , U = T
+  > =
+    & ( ( setState: A.Get<Mutate<Store<T>, Mis>, "setState", undefined>
+        , getState: A.Get<Mutate<Store<T>, Mis>, "getState", undefined>
+        , store: Mutate<Store<T>, Mis>
+        , $$storeMutations: Mis
+        ) =>
+          U
       )
-    }
-    listeners.add(listener as StateListener<TState>)
-    // Unsubscribe
-    return () => listeners.delete(listener as StateListener<TState>)
+    & { [$$storeMutators]?: Mos }
+declare const $$storeMutators: unique symbol;
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+interface StoreMutators<S, A> {}
+type StoreMutatorIdentifier = keyof StoreMutators<unknown, unknown>
+
+type Mutate<S, Ms> =
+  Ms extends [] ? S :
+  Ms extends [[infer Mi, infer Ma], ...infer Mrs]
+    ? Mutate<
+        StoreMutators<S, Ma>[Mi & StoreMutatorIdentifier],
+        Mrs
+      > :
+  never
+
+
+
+
+// ============================================================================
+// Implementation via Existential Types
+
+type EState = { __isState: true }
+type EShouldReplaceState = boolean & { __isShouldReplaceState: true }
+
+interface EStore
+  { getState:
+      () => EState
+  , setState:
+      ( nextStateOrUpdater:
+          | O.Partial<EState>
+          | ((state: EState) => O.Partial<EState>)
+      , shouldReplace?: EShouldReplaceState
+      ) =>
+        void
+  , subscribe:
+      ( listener:
+          (state: EState, previousState: E.Previous<EState>) => void
+      ) =>
+        () => void
+  , destroy:
+      () => void
+  }
+
+type ECreate =
+  ( storeInitializer:
+    ( set: EStore['setState']
+    , get: EStore['getState']
+    , store: EStore
+    ) => EState
+  ) =>
+    EStore
+
+const createImpl: ECreate = storeInitializer => {
+  let state: EState
+
+  type Listener = (state: EState, previousState: E.Previous<EState>) => void
+  const listeners: Set<Listener> = new Set()
+  const emit = (...a: F.Arguments<Listener>) =>
+    listeners.forEach(f => f(...a))
+
+  const store: EStore = {
+    getState: () => state,
+    setState: (nextStateOrUpdater, shouldReplace) => {
+      const nextState =
+        typeof nextStateOrUpdater === 'function'
+          ? nextStateOrUpdater(state)
+          : nextStateOrUpdater
+
+      if (objectIs(nextState, state)) return
+
+      const previousState = E.previous(state)
+      state = shouldReplace
+        ? nextState as EState
+        : Object.assign({}, state, nextState)
+
+      emit(state, previousState)
+    },
+    subscribe: listener => {
+      listeners.add(listener)
+      return () => listeners.delete(listener)
+    },
+    destroy: () => listeners.clear()
   }
+  state = storeInitializer(store.setState, store.getState, store)
+  return store
+}
+const create = createImpl as unknown as Create
+
+
+
+
+// ============================================================================
+// Utilities
+
+const objectIs = Object.is as (<T>(a: T, b: T) => boolean)
+
+namespace E {
```

Comment:
I was planning to request some small things later, but maybe it's a good timing. For a start, can we avoid using namespace? We strongly prefer explicitly named type aliases.

Reply:
Sure no problems, I'll avoid namespaces.

I usually use them to create inlined modules that work on a specific abstraction or type, you can read more [here](https://twitter.com/devanshj__/status/1427800465169731585). But there's not much utility in this scenario anyway.